text/3476-expose-fn-type.md

```diff
@@ -0,0 +1,164 @@
+
+For that, I would need to mark the type behind the function, for example, with a trait.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As we all know, you can refer to a struct by its name and for example implement a trait
+```rust
+struct Timmy;
+impl Person for Timmy {
+    fn greet() {
+        println!("Hey it's me, Timmy!");
+    }
+}
+```
+When we want to target a specific function for a trait implementation, we somehow need to get to the type behind it. That is being done with the `fn` keyword as follows
+```rust
+fn my_function() {}
+impl MyTrait for fn my_function {
+    /* ... */
+}
+```
+---
+For a better understanding, imagine you have a struct like this:
+```rust
+struct FnContainer<F: Fn()> {
+    inner: F,
+}
+fn goods() { }
+
+let contained_goods = FnContainer {
+    inner: goods
+};
+```
+Here, we make a `FnContainer` which can hold every function with the signature `() -> ()` via generics.
+But what about explicitly designing the `FnContainer` for a specific function, just like the compiler does when resolving the generics. This will work the same as with the trait impl from above:
+```rust
+struct GoodsContainer {
+    inner: fn goods,
+}
+fn goods() {}
+
+let contained_goods = GoodsContainer {
+    inner: goods,
+}
+```
+---
+A function with a more complex signature, like with parameters, modifiers or a return type, is still just referenced by its name, because it's already unique
+```rust
+async fn request_name(id: PersonID) -> String { .. }
+
+impl Requestable for fn request_name {
+    /* ... */
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As described in the **Guide-level explanation**, with the syntax `fn <fn_path>`, we can reference the type behind the named function.
+
+When the function is for example in a different mod, it should be referenced by its path
+```rust
+mod sub {
+    fn sub_mod_fn() { .. }
+}
+impl fn sub::sub_mod_fn {
+    /* ... */
+}
+```
+
+It should be also possible to get the type of functions inside impl blocks:
+
+```rust
+struct MyStruct;
+impl MyStruct {
+    fn new() -> Self { Self }
+}
+impl fn MyStruct::new {
+    /* ... */
+}
+```
+
+When a function has generics, they will be handled as follows, just like we know it from normal types
+```rust
+fn send<T: Send>(val: T) {}
+impl<T: Send> ParcelStation for fn send<T> {
+    /* ... */
+}
+```
+
+When we have an implicit generic, they will be appended in order to the generic list:
+```rust
+fn implicit_generic(val: impl Clone) -> impl ToString {}
+impl<T: Send, U: ToString> for fn implicit_generic<T, U> {
+    /* ... */
+}
+```
+
+Just as structs and enums have the possibility to derive traits to automatically generate code, function type do too
```

Comment:
This may be a major problem for macro developers. For one thing, `syn::DeriveInput` currently has no `Verbatim` variant, so an extra derive input type will most likely be impossible to implement without introducing syn v3 (oh no, we just got (not yet) everyone upgraded to syn v2).

Plus, I don't see any practical reason to need to derive traits for specific fn pointers (which are always ZST) other than for `Debug`, which we can already implement in libcore instead. If traits defined by libraries have a specific requirement to implement on specific function pointers that cannot be done generically, they can use an attribute macro instead of a derive macro for this specific case.

Reply:
That's the feedback i was waiting for 😅. I wasn't sure how easily it will be so i threw it into the RFC and waited to see if anyone would complain. So back to attribute macros